include/fiction/io/print_layout.hpp

```diff
@@ -343,41 +353,147 @@ void print_charge_layout(std::ostream& os, const Lyt& lyt, const bool cs_color =
+        }
     }
 
-    coordinate<Lyt> min{};
-    coordinate<Lyt> max{lyt.x(), lyt.y(), 1};
+    // determine the north-west and south-west cell in siqad-coordinates.
+    const auto [nw, se] = bounding_box_siqad<Lyt>(lyt);
 
-    if (crop_layout)
+    auto min_nw = nw;
+    auto max_se = se;
+
+    std::vector<typename Lyt::cell> defects{};
+
+    // Check if defects exist in the layout.
+    if constexpr (has_get_sidb_defect_v<Lyt>)
     {
-        const auto bb = bounding_box_2d{lyt};
 
+        defects.reserve(lyt.num_defects());
+        lyt.foreach_sidb_defect([&defects](const auto& c) { defects.push_back(c.first); });
+        std::sort(defects.begin(), defects.end());
+        if (!defects.empty())
+        {
+            min_nw = min_nw > defects.front() ?
+                         defects.front() :
+                         min_nw;  // if a defect is more north-west than nw, this position is used as min.
+            max_se = max_se < defects.back() ?
+                         defects.back() :
+                         max_se;  // if a defect is more south-east than se, this position is used as max.
+        }
+    }
+
+    if (crop_layout)
+    {
         // apply padding of maximally one dimer row and two columns
-        min = bb.get_min() - coordinate<Lyt>{2, 1};
-        max = bb.get_max() + coordinate<Lyt>{2, 1};
+        min_nw = min_nw - coordinate<Lyt>{2, 1};
+        max_se = max_se + coordinate<Lyt>{2, 1};
 
         // ensure only full dimer rows are printed
-        min.z = 0;
-        max.z = 1;
+        min_nw.z = 0;
+        max_se.z = 1;
     }
 
-    // iterate over all coordinates in the rows determined by the vertical crop
-    lyt.foreach_coordinate(
-        [&](const coordinate<Lyt>& c)
+    // loop coordinate is initialized with the north-west coordinate.
+    auto loop_coordinate = min_nw;
+
+    while (loop_coordinate <= max_se)
+    {
+        // Check if layout is a sidb_surface and a charge distribution surface.
+        if constexpr (has_get_sidb_defect_v<Lyt> && has_get_charge_state_v<Lyt>)
         {
-            if (crop_layout && (c.x < min.x || c.x > max.x))  // apply horizontal crop
+            if (lyt.get_sidb_defect(loop_coordinate) != sidb_defect{sidb_defect_type::NONE})
             {
-                return;
+                if (is_negatively_charged_defect(lyt.get_sidb_defect(loop_coordinate)))
+                {
+                    os << fmt::format(cs_color ? detail::SIDB_NEG_COLOR : detail::NO_COLOR, " ⊟ ");
+                }
+
+                else if (is_positively_charged_defect(lyt.get_sidb_defect(loop_coordinate)))
+                {
+                    os << fmt::format(cs_color ? detail::SIDB_NEG_COLOR : detail::NO_COLOR, " ⊞ ");
+                }
+                else if (is_neutrally_charged_defect(lyt.get_sidb_defect(loop_coordinate)))
+                {
+                    os << fmt::format(cs_color ? detail::SIDB_NEG_COLOR : detail::NO_COLOR, " ⊡ ");
+                }
```

Comment:
These blocks (the following ones as well) contain a lot of code duplication. Please consider using a more concise representation by checking individually for the presence of charges or defects.